aten/src/ATen/native/UnaryOps.cpp

```diff
@@ -31,6 +31,122 @@
 namespace at {
 namespace native {
 
+namespace {
+  enum class TypePromotionStrategy {
+    None,  // No implicit dtype promotion
+    Type1, // int8 - fp16, int16 - fp32, int32 - fp64, int64 - fp64, bool - fp16
+    Type2, // (int8, int16, int32, int64, bool) - fp64
+    Type3, // bool - int8
+    Type4  // bool - fp16
+  };
+
+  // promoteToFloatType[1/2/3/4] functions are helper functions to support input dtype to Float implicit promotions 
+  // based on NumPy's conversion rules
+  // For discussion, check https://github.com/pytorch/pytorch/pull/33322 and https://github.com/pytorch/pytorch/issues/28703
+  // There are 4 type of dtype promotions in NumPy: (float16 is replaced by float32 for CPU devices)
+  // Type - 1: int8 - float16, int16 - float32, int32 - float64, int64 - float64, bool - float16
+  // Type - 2: (int8, int16, int32, int64, bool) - float64
+  // Type - 3: bool - int8
+  // Type - 4: bool - float16
+  static inline ScalarType promoteToFloatType1(const Tensor& self) {
+    // This promotes dtype based on Type 1 strategy
+    ScalarType dtype;
+    switch(self.scalar_type()) {
+      case kChar:
+        dtype = (self.device().type() == DeviceType::CPU) ? kFloat : kHalf;
+        break;
+      case kShort:
+        dtype = kFloat;
```

Comment:
might be simpler to just put the return statements in each case here instead of saving and breaking to return later.

Reply:
Agreed. This should be resolved here: https://github.com/pytorch/pytorch/pull/33322/commits/2203eb9fae716bad2b5907dfc4518323ca4ee5c5